server/app/controllers/admin/AdminProgramStatusesController.java

```diff
@@ -52,215 +51,131 @@ public AdminProgramStatusesController(
-      throws MissingStatusException, DuplicateStatusException {
-    ImmutableMap<String, Integer> statusNameToIndex = statusNameToIndexMap(statuses);
-    if (!statusNameToIndex.containsKey(replaceStatusName)) {
-      throw new MissingStatusException();
-    }
-    StatusDefinitions.Status editedStatus =
-        replacerFunc.apply(statuses.get(statusNameToIndex.get(replaceStatusName)));
-    if (!replaceStatusName.equals(editedStatus.statusText())
-        && statusNameToIndex.containsKey(editedStatus.statusText())) {
-      throw new DuplicateStatusException(editedStatus.statusText());
+  /**
+   * Deletes a status that is currently associated with the program as signified by the
+   * "deleteStatusText" form parameter.
+   */
+  @Secure(authorizers = Authorizers.Labels.CIVIFORM_ADMIN)
+  public Result delete(Http.Request request, long programId) throws ProgramNotFoundException {
+    if (!statusTrackingEnabled) {
+      return notFound("status tracking is not enabled");
     }
-    return IntStream.range(0, statuses.size())
-        .boxed()
-        .map(
-            i -> {
-              if (i.equals(statusNameToIndex.get(replaceStatusName))) {
-                return editedStatus;
-              }
-              return statuses.get(i);
-            })
-        .collect(ImmutableList.toImmutableList());
-  }
+    requestChecker.throwIfProgramNotDraft(programId);
+    ProgramDefinition program = service.getProgramDefinition(programId);
 
-  private static ImmutableList<StatusDefinitions.Status> removeStatus(
-      ImmutableList<StatusDefinitions.Status> statuses, String removeStatusName)
-      throws MissingStatusException {
-    ImmutableMap<String, Integer> statusNameToIndex = statusNameToIndexMap(statuses);
-    if (!statusNameToIndex.containsKey(removeStatusName)) {
-      throw new MissingStatusException();
+    DynamicForm requestData = formFactory.form().bindFromRequest(request);
+    String rawDeleteStatusText = requestData.get(ProgramStatusesView.DELETE_STATUS_TEXT_NAME);
+    if (Strings.isNullOrEmpty(rawDeleteStatusText)) {
+      return badRequest("missing or empty status text");
     }
-    return IntStream.range(0, statuses.size())
-        .boxed()
-        .filter(i -> !i.equals(statusNameToIndex.get(removeStatusName)))
-        .map(i -> statuses.get(i))
-        .collect(ImmutableList.toImmutableList());
-  }
-
-  private static ImmutableMap<String, Integer> statusNameToIndexMap(
-      ImmutableList<StatusDefinitions.Status> statuses) {
-    return IntStream.range(0, statuses.size())
-        .boxed()
-        .collect(
-            ImmutableMap.toImmutableMap(i -> statuses.get(i).statusText(), i -> i, (s1, s2) -> s1));
-  }
-
-  private static final class MissingStatusException extends Exception {}
-
-  private static final class DuplicateStatusException extends Exception {
-    final String userFacingMessage;
+    final String deleteStatusText = rawDeleteStatusText.trim();
```

Comment:
Good catch. Agreed with "sneaking in as a real status". We likely could move to a model where we trim the status entered by the user in the form before serializing, but always use the raw value from the DB in the relevant fields. That would allow us to still perform the operation for "bad" data while having the controllers try to put the "right" value in.

I left a TODO for this since it's pretty low priority in my opinion and I'd like to iterate on smaller PRs once the initial functionality is in (if that's acceptable from your end).